### PR TITLE
Added checks for volume files

### PR DESCRIPTION
**Pull request by aonez**
From pull https://github.com/abbeycode/UnrarKit/pull/36

This one contains various functions to check wether the file is a parted RAR volume, the get the first part path and to get the list of all the parts paths of the volume.

### CODE REVIEW

**Comment by aonez** [comment]
Here a sample code using some of the functions:

```objective-c
NSError *archiveError = nil;
URKArchive *archive = [[URKArchive alloc] initWithPath:rar error:&archiveError];
NSError *error = nil;
        
NSArray<URKFileInfo*> *fileInfosInArchive = [archive listFileInfo:&error];
        
if (archive.isVolume) {
	NSLog(@"isVolume");
	NSLog(@"First volume: %@", [archive firstVolumePath]);
            
	URKArchive *volarchive = [[URKArchive alloc] initWithPath:[archive firstVolumePath] error:&archiveError];
	NSArray<NSString*> *paths = [volarchive listVolumePaths:&error];
	NSLog(@"List of volumes:");

	for (NSString *path in paths) {
		NSLog(@"%@", path);
	}
            
}
else
	NSLog(@"NOT isVolume");
```

**Comment by aonez** [comment]
Just noticed the code does not detect the volume parts if called with the first part. (.rar or part1.rar). I'll update the pull as soon as I fix it.

**Comment by abbeycode** [comment]
Thanks for keeping up on this! It looks like it'll make a good addition to the library. I'll review as soon as I'm able to.

**Comment by abbeycode** [comment]
I'm starting to take a look at this now. To give me some context, can you explain how your code is using multi-volume archives, and how each of the methods you defined fits into your use case?

**Comment by aonez** [comment]
I needed a way to list all the parts of a volume, so I can delete them all after an extraction. Also it is useful when a user tries to extract a volume using all the parts, I can select the first one and dismiss the others.

First, this one tells me if the file is a volume part, so I can search for the first one and list all the parts. I'll try `0x0010` flag to detect the first volume (thanks for the technote tip):
```objective-c
- (BOOL)isVolume:(NSURL *)fileURL
```

If it is a volume, I want to be sure it is the first part. With this function I'll get the first volume path:
```objective-c
- (NSString *)firstVolumePath:(NSString *)filePath
```

Then, I can just extract the file using the first path. Also, I can get all the volume paths from the first one, so I can delete them or dismiss them from the extraction process: 
```objective-c
- (NSArray<NSString*> *)listVolumePaths:(NSError **)error
```

This last function maybe can be done in another way. Right now I list the contents of the file so I can get the paths of all (found) volumes. Maybe it can be done using `0x0008` and `0x0010` flags and wishing the previous/next volume path and checking for it.

**Comment by aonez** [comment]
I'll be using it here: https://github.com/aonez/Keka/issues/8

**Comment by aonez** [comment]
Checked again, maybe I can't get the volume info in the header because you're using `RARHeaderDataEx` instead of `RARHeaderData`.

**Comment by abbeycode** [comment]
Sorry it took me a while to get back to you, but I was getting the code all cleaned up to work with the latest Xcode version, and it took longer than expected. That's great you found a better way to list the volume paths! Can you push an updated commit?

Also, I'm curious about why you need the `- (BOOL)isVolume:(NSURL *)fileURL` and `- (NSString *)firstVolumePath:(NSString *)filePath` overloads. When are the simpler `- (BOOL)isVolume` and `- (NSString *)firstVolumePath` not sufficient?

I'll try to keep up on this a little more to get it out the door. I'm going to be pushing a branch I created that'll help us get started on unit tests for these new methods. I'm using the `rar` CLI tool to create some multi-volume archives and run tests on them. I'll post more when I'm ready for you to look at them and integrate them into your PR.

**Comment by aonez** [comment]
@abbeycode not hurry at all. Sadly I did not found any other way to get the first part of the volume detected as a volume, nor get a consistent way to detect the next/previous volume existence. So right now the code is as here.

The overloads are not necessary for me. I'm using with the opened archive (`self.filename`), but I think maybe someone will need to pass another path, and why not coding it already this capability. If you like I can change it to use always `self.filename` and get rid of the path overload, no problem at all.

Let me know how the tests go! Maybe you get motivated and find the right flag to get the first volume path detected ;)

**Comment by aonez** [comment]
Just rebased with your master, seems there're too many changes now in the pull request...

**Comment by abbeycode** [comment]
That's strange that it's showing all of those as changes. Did you do a `git rebase`, or did you `git merge`?

**Comment by aonez** [comment]
Did a rebase but using the [Fork](https://git-fork.com) GUI. First rebase for me by the way, so not surprised it got broken...

**Comment by abbeycode** [comment]
Interesting... I think it messed it up. Also, regardless of what it did, I think GitHub expects you to just `merge`, not `rebase`. I've run into it before. Can you try doing a hard reset of your branch to a previous revision and do it over?

**Comment by aonez** [comment]
Sure, I’ll try to fix it right away :)

**Comment by aonez** [comment]
Done! Hard reset and a merge did it.

**Comment by abbeycode** [comment]
I added the tests and stubbed out methods. I renamed the one method to be a little clearer, and added `NSURL`-based overloads and documentation.

Also, while I was running the tests with your current implementation, it looks like you're going to want to use this in your `isVolume` check:

    bool isVolume = (flags->Flags & MHD_VOLUME) != 0;

`RARReadHeaderEx()` reads file headers, not the archive header.

Let me know if you have any other questions! Also, you should probably update the PR to merge into that new `multivolume` branch I created instead of `master`. I'll merge it in from there.

**Comment by aonez** [comment]
You got it, using `flags` the first volume is detected. I'll spend some time next week to merge it into the branch and take a look at all the test and modifications you made. Thanks @abbeycode!

**Comment by aonez** [comment]
Little busy right now, but I'll get to it!

**Comment by abbeycode** [comment]
Hello. It's been a while, and I was wondering if you were still planning on pushing this PR over the finish line.

**Comment by aonez** [comment]
Just was thinking about this some days ago. Sure I will!

**Comment by aonez** [comment]
Ok, so my PR is updated to the current status of the `multivolume` branch. Let me know when you update it with `v2.9`changes so I'll update it again 😅

**Comment by abbeycode** [comment]
I updated the `multivolume` branch. Let's make sure the [Travis build](https://travis-ci.org/abbeycode/UnrarKit/builds/265611037) completes successfully before you merge. I'll send another note when it looks good.

**Comment by abbeycode** [comment]
The build failed, but that's because of the unit tests that this PR will fix, so you may proceed!

**Comment by aonez** [comment]
Should work now, let's see what travis says 😛

**Comment by abbeycode** [comment]
Hmmmm still looks like the new unit tests are failing...

**Comment by aonez** [comment]
I'll take a look in a few hours...

**Comment by aonez** [comment]
Cleaned all my code and failed again. Then I saw your branch failed too. Look at your latest commit:

https://github.com/abbeycode/UnrarKit/commits/multivolume

**Comment by abbeycode** [comment]
Sorry, I think the PR is all messed up. It mostly looks like my v2.9 branch, with none of your own changes. I think it went wrong when you rebased – try merging instead.

And remember after you get it sorted out, Travis CI is just running the unit tests that are in the repo. Before you push your commits next time, try running all of the tests in Xcode (for the Mac framework, which is where the multivolume tests run) and see if they pass. If they don't, figure out why.

**Comment by aonez** [comment]
I've made a rebase with the `master` instead of `multivolume` 😨

Anyway I'll take a look at your test. For starters I will add the test files the methods are trying to open... Hope to finish this today 👍

**Comment by aonez** [comment]
I've passed all volume related tests on my machine...

- [x] FirstVolumeTests
- [x] HasMultipleVolumesTests
- [x] ListVolumesTests

I had some trouble on the `listVolumePaths` method. If the current archive is not the first, `listFileInfo` only returns the current and the next volumes names. So the correct thing here is to check if the current is the first and if not change to it. I've used this one, that works, but I'm not sure I'm supposed to use that one from inside `URKArchive.m`:

```objective-c
[self initWithURL:firstVolumeURL error:&firstVolumeURLError];
```

Because of travis CI, had to assign that call to a variable (`dummy`) and use it, so:

```objective-c
URKArchive * dummy = self;
...
dummy = [self initWithURL:firstVolumeURL error:&firstVolumeURLError];
...
NSArray<URKFileInfo*> *listFileInfo = [dummy listFileInfo:error];
```

When it can be (and actually works and passes the tests on my machine):
```objective-c
[self initWithURL:firstVolumeURL error:&firstVolumeURLError];
...
NSArray<URKFileInfo*> *listFileInfo = [self listFileInfo:error];
```

Just pointing it to you, maybe you have a more clever way to approach this one.

**Comment by aonez** [comment]
I can't pass all the test on my machine because `testExtractBufferedData_VeryLarge` gives throws an `EXC_BAD_ACCESS`

File: Classes/URKArchive.mm [hunk 1]
```diff
@@ -331,7 +403,32 @@ - (nullable NSURL *)firstVolumeURL {
...
-    return @[];
+    __block NSMutableArray<NSString*> *volumePaths = [NSMutableArray new];
+    
+    URKArchive * dummy = self;
+    NSURL * firstVolumeURL = [self firstVolumeURL];
+    if (firstVolumeURL != self.fileURL)
+    {
+        NSError *firstVolumeURLError = nil;
+        
+        dummy = [self initWithURL:firstVolumeURL error:&firstVolumeURLError];
```
Comment (abbeycode) [inline Classes/URKArchive.mm]: We need to find a different solution here. You should never call [self init...] from a Cocoa class. What error are you checking for? This only returns an error when it can’t create a bookmark. What do you need to create the bookmark over again for?

File: Classes/URKArchive.mm [hunk 2]
```diff
@@ -331,7 +403,32 @@ - (nullable NSURL *)firstVolumeURL {
 
 - (nullable NSArray<NSString*> *)listVolumePaths:(NSError **)error
 {
-    return @[];
+    __block NSMutableArray<NSString*> *volumePaths = [NSMutableArray new];
```
Comment (abbeycode) [inline Classes/URKArchive.mm]: This doesn’t need to be marked as a block variable if you’re not setting it from within a block. Even adding members to a mutable collection can be done without __block.

File: Classes/URKArchive.mm [hunk 3]
```diff
@@ -331,7 +403,32 @@ - (nullable NSURL *)firstVolumeURL {
...
+        }
+    }
+    
+    NSArray<URKFileInfo*> *listFileInfo = [dummy listFileInfo:error];
+    
+    if (listFileInfo == nil)
+        return nil;
+    
+    for (URKFileInfo* info in listFileInfo) {
+        if (![volumePaths containsObject:info.archiveName])
```
Comment (abbeycode) [inline Classes/URKArchive.mm]: Rather than performing this check every time, make the collection an NSMutableSet instead, so it only contains unique members.

File: Classes/URKArchive.h [hunk 4]
```diff
@@ -234,6 +234,22 @@ extern NSString *URKErrorDomain;
 + (BOOL)urlIsARAR:(NSURL *)fileURL;
 
 /**
+ *  Determines whether the current file is a RAR volume part
```
Comment (aonez) [inline Classes/URKArchive.h]: Did not get why you've added `hasMultipleVolumes`. Removing those 👍

File: Classes/URKArchive.mm [hunk 5]
```diff
@@ -315,8 +351,44 @@ + (BOOL)urlIsARAR:(NSURL *)fileURL
     return [NSArray arrayWithArray:fileInfos];
 }
 
+- (nullable NSString *)firstVolumePath:(NSString *)filePath
```
Comment (aonez) [inline Classes/URKArchive.mm]: The idea of the overloads (same to previous comment) was to make this methods useful for checking a file that is not the one loaded during the init process. For example, while searching for the first volume and checking if it is in fact a volume. I'm gonna remove them, so It'll be always checking for `self.fileURL`.

File: Classes/URKArchive.h (hunk 4 above)
Comment (aonez) [inline Classes/URKArchive.h]: Now, I'm thinking, since the name implied multiple, I've made `hasMultipleVolumes` to check if it actually has ***multiple*** volume (checking `listVolumeURLs`)... Now I see you mean if it is a part of a volume, even if it is left alone and no other parts of the volume are to be found (extraction will fail, obviously), so I'll just check `isVolume`.

Comment (aonez) [inline Classes/URKArchive.h]: Or you mean moving the code from `isVolume` to `hasMultipleVolumes`? 🤔

File: Classes/URKArchive.mm (hunk 1 above)
Comment (aonez) [inline Classes/URKArchive.mm]: In `listVolumePaths` we need to return all the volumes available. If, for example, I've created the `URKArchive` instance with volume 3 of 4, when checking `listFileInfo` it will only return 3 and 4. To get 1-3, self.fileURL need to be the volume 1. So this pretty piece of code did that, changing self.fileURL to `firstVolumeURL`, when needed.

That's why I told you about the overload with `fileURL`, so I can use some methods with other files than `self.fileURL`.

File: Classes/URKArchive.mm (hunk 3 above)
Comment (aonez) [inline Classes/URKArchive.mm]: Since `NSMutableSet` is unordered, I'll edit the `testMultipleVolume_UseFirstVolume`, since it fails because of the array order.

File: Classes/URKArchive.mm (hunk 1 above)
Comment (aonez) [inline Classes/URKArchive.mm]: I'm going with another approach 👍

File: Classes/URKArchive.h (hunk 4 above)
Comment (aonez) [inline Classes/URKArchive.h]: Just to be clear, can't use `isVolume` because it does not detect as a volume the first part. I'm removing this method.

https://github.com/abbeycode/UnrarKit/pull/38#issuecomment-277518409
> ...Sadly I did not found any other way to get the first part of the volume detected as a volume, nor get a consistent way to detect the next/previous volume existence. So right now the code is as here.

**Comment by aonez** [comment]
Ok, I'll relax a bit now, still saturday 😂

File: Classes/URKArchive.mm (hunk 3 above)
Comment (abbeycode) [inline Classes/URKArchive.mm]: It would probably be a better idea to sort the list when you convert to an array.

Comment (aonez) [inline Classes/URKArchive.mm]: Would check it out in two weeks. Will be busy until then. Let me know if there's something else to change :)

File: Classes/URKArchive.mm [hunk 6]
```diff
@@ -331,7 +371,61 @@ - (nullable NSURL *)firstVolumeURL {
...
+        if (listFileInfo == nil) {
+            return nil;
+        }
+        
+        for (URKFileInfo* info in listFileInfo) {
+            [volumePaths addObject:info.archiveName];
+        }
+    }
+
+    return [NSArray arrayWithArray:volumePaths.allObjects];
```
Comment (abbeycode) [inline Classes/URKArchive.mm]: Sort by path here for predictable behavior, and then you won't need to modify the unit test cases.

File: Classes/URKArchive.mm [hunk 7]
```diff
@@ -331,7 +371,61 @@ - (nullable NSURL *)firstVolumeURL {
...
-    return @[];
+    NSMutableSet<NSString*> *volumePaths = [NSMutableSet new];
+    
+    NSURL * firstVolumeURL = [self firstVolumeURL];
+    if (firstVolumeURL != self.fileURL)
+    {
+        @try {
+            NSError *firstVolumeError = nil;
+            
+            if ([self _unrarOpenFile:firstVolumeURL.path
```
Comment (abbeycode) [inline Classes/URKArchive.mm]: This is only a little bit better than calling `[self init...]`. Calling this method changes private members, that now correspond to a different archive's file. How about instead of using `self` at all, make a new `URKArchive` instance here using `firstVolumeURL`? You can still call private methods on it. Perhaps make another private method named `searchForVolumePaths` (or something like that) that returns the list of volume paths. Use `-listFileInfo:` as an example. You can just return the result of that method on the `firstVolumeArchive` instance instead of this entire `@try` block.

Another approach could be to only ever use the first volume in `init...`. If a user initializes an archive with an extension that contains a number, replace the filepath with a path to the first volume. You can refactor the guts of `firstVolumePath` to be a class method that takes a path as an argument, and call that from the instance method with `self.filename` as its argument. You would do those checks before any member variables get assigned.

What do you think?

File: Tests/ListVolumesTests.m [hunk 8]
```diff
@@ -61,7 +61,7 @@ - (void)testMultipleVolume_UseFirstVolume {
     XCTAssertNil(listVolumesError, @"Error listing volume paths");
     XCTAssertNotNil(volumePaths, @"No paths returned");
     XCTAssertEqual(volumePaths.count, 5, @"Wrong number of volume paths listed");
-    XCTAssertTrue([expectedVolumePaths isEqualToArray:volumePaths],
+    XCTAssertTrue([[NSSet setWithArray:expectedVolumePaths] isEqual:[NSSet setWithArray:volumePaths]],
```
Comment (abbeycode) [inline Tests/ListVolumesTests.m]: As mentioned, if you sort the results of `listVolumePaths` then you can revert these changes to the unit tests.

**Comment by aonez** [comment]
Getting with this next week! Really busy lately 😅

**Comment by abbeycode** [comment]
That's fine, don't sweat it. When you do have time though, I've updated the v2.9 branch since you last merged. Please sync the upstream changes.

**Comment by abbeycode** [comment]
I'm ready when you are :-) As mentioned before, just make sure you sync to the updates on the `multivolume` branch. As an added bonus, I've fixed the validation scripts, so they should work correctly for pull requests now :)

File: Classes/URKArchive.mm (hunk 6 above)
Comment (aonez) [reply Classes/URKArchive.mm]: Going with NSMutableOrderedSet

**Comment by aonez** [comment]
> Another approach could be to only ever use the first volume in init.... If a user initializes an archive with an extension that contains a number...

I like a loot this idea. But maybe a init overload like `- (instancetype)initWithPath:(NSString *)filePath checkFirstVolume:(bool)checkFirstVolume error:(NSError **)error` so the developer can decide wether to use this feature?

File: Classes/URKArchive.mm [hunk 9]
```diff
@@ -255,7 +255,15 @@ - (NSNumber *)compressedSize
...
-    return NO;
+    NSError *listError = nil;
+    NSArray<NSURL*> *volumeURLs = [self listVolumeURLs:&listError];
+    
+    if (!volumeURLs) {
+        NSLog(@"Error getting file volumes list: %@", listError);
+        return false;
+    }
+    
+    return (volumeURLs.count > 1);
```
Comment (abbeycode) [inline Classes/URKArchive.mm]: _Very very minor suggestion_: No need for the parentheses here.

File: Classes/URKArchive.mm (hunk 7 above)
Comment (abbeycode) [reply Classes/URKArchive.mm]: Please take a look at this suggestion. It's the only real thing holding this back. Thanks for hopping back on this request!

Comment (abbeycode) [reply Classes/URKArchive.mm]: It looks like you actually did reply to this (I'm seeing it in an email, but not on the page).

> I like a loot this idea. But maybe a init overload like - (instancetype)initWithPath:(NSString *)filePath checkFirstVolume:(bool)checkFirstVolume error:(NSError **)error so the developer can decide wether to use this feature?

I can think of two reasons to make it a separate overload:

1. The old behavior is desirable for some use cases
2. In case the new version has a bug

Regarding 1, I can't think of any reason to ever operate on a subsequent volume instead of the original. I think this is how most RAR extractors work. Can you think of any?

2 may possibly be a concern, but we should make sure the updated `init` gracefully falls back to the existing behavior, logging as it goes to help a developer figure out what may have happened. Basically, do and log each of the following:

* Look for a new root volume if it looks like a subsequent part of a multivolume archive
* If yes to the above, look for the root archive
* If not found, continue as if the one given is the root
* If found, substitute the given path for the root volume path

Thoughts?

Comment (aonez) [reply Classes/URKArchive.mm]: I did replied... but then messed with the reply...

1. Thought about volume reconstruction, in case the first part does not exist. But with the code as is right now, it will not return a first part if it does not exist, so no case that I can think here.
2. Maybe just adding the correct logs on the first volume check will do, as you say.

File: Classes/URKArchive.h [hunk 10]
```diff
@@ -289,16 +289,20 @@ extern NSString *URKErrorDomain;
...
 
 /**
  *  Gets the URL of the first volume in a multi-volume archive, or the only volume in a single-volume one
  *
+ *  @param fileURL  The URL of the archive file
+ *
  *  @return Returns the URL, or nil if there was an error
  */
-- (nullable NSURL *)firstVolumeURL;
+- (NSURL *)firstVolumeURL:(NSURL *)fileURL;
```
Comment (aonez) [inline Classes/URKArchive.h]: Those will be not nullable, since they will be called after a (!fileURL) check. If no first part, they will return the passed one.

File: Classes/URKArchive.mm [hunk 11]
```diff
@@ -255,7 +257,15 @@ - (NSNumber *)compressedSize
...
-    return NO;
+    NSError *listError = nil;
+    NSArray<NSURL*> *volumeURLs = [self listVolumeURLs:&listError];
+    
+    if (!volumeURLs) {
+        NSLog(@"Error getting file volumes list: %@", listError);
+        return false;
+    }
+    
+    return volumeURLs.count > 1;
```
Comment (aonez) [inline Classes/URKArchive.mm]: No extra parenthesis 👍

File: Classes/URKArchive.mm [hunk 12]
```diff
@@ -385,23 +395,75 @@ + (BOOL)urlIsARAR:(NSURL *)fileURL
...
-    return @"";
+- (NSString *)firstVolumePath:(NSString *)filePath {
+    NSURL *fileURL = [self firstVolumeURL:[NSURL fileURLWithPath:filePath]];
+
+    return fileURL.path;
 }
 
-- (nullable NSURL *)firstVolumeURL {
-    NSString *path = [self firstVolumePath];
+- (NSURL *)firstVolumeURL:(NSURL *)fileURL {
```
Comment (aonez) [inline Classes/URKArchive.mm]: Changed the contents to the URL method because `init` uses URL

File: Classes/URKArchive.mm [hunk 13]
```diff
@@ -385,23 +395,75 @@ + (BOOL)urlIsARAR:(NSURL *)fileURL
...
     }
     
-    return [NSURL fileURLWithPath:path];
+    if (match) {
+        if ([[NSFileManager defaultManager] fileExistsAtPath:volumePath]) {
+            URKLogDebug("First volume part %@ found. Using as the main archive", volumePath);
+            return [NSURL fileURLWithPath:volumePath];
+        }
+        else
+            URKLogInfo("First volume part %@ not found. Skipping first volume selection", volumePath);
```
Comment (aonez) [inline Classes/URKArchive.mm]: Just skip, do not return nil

File: Classes/URKArchive.mm [hunk 14]
```diff
@@ -385,23 +395,75 @@ + (BOOL)urlIsARAR:(NSURL *)fileURL
...
+    }
+
+    return fileURL;
```
Comment (aonez) [inline Classes/URKArchive.mm]: If no match, always return the passed fileURL

File: Classes/URKArchive.mm [hunk 15]
```diff
@@ -385,23 +395,75 @@ + (BOOL)urlIsARAR:(NSURL *)fileURL
...
 }
 
 - (nullable NSArray<NSString*> *)listVolumePaths:(NSError **)error
 {
-    return @[];
+    NSMutableOrderedSet<NSString*> *volumePaths = [NSMutableOrderedSet new];
+
+    NSArray<URKFileInfo*> *listFileInfo = [self listFileInfo:error];
```
Comment (aonez) [inline Classes/URKArchive.mm]: Way cleaner 😃

File: Classes/URKArchive.mm [hunk 16]
```diff
@@ -134,7 +134,9 @@ - (instancetype)initWithFile:(NSURL *)fileURL password:(NSString*)password error
         if (error) {
             *error = nil;
         }
-
+        
+        fileURL = [self firstVolumeURL:fileURL];
```
Comment (aonez) [inline Classes/URKArchive.mm]: Just added it before `(!fileURL)` check and `[self init]`, let me know if you think there's a better place

File: Classes/URKArchive.mm [hunk 17]
```diff
@@ -385,23 +395,75 @@ + (BOOL)urlIsARAR:(NSURL *)fileURL
...
+    return fileURL.path;
 }
 
-- (nullable NSURL *)firstVolumeURL {
-    NSString *path = [self firstVolumePath];
+- (NSURL *)firstVolumeURL:(NSURL *)fileURL {
     
-    if (![path length]) {
-        return nil;
+    URKLogDebug("Checking if the archive is part of a volume...");
```
Comment (aonez) [inline Classes/URKArchive.mm]: Added `URKLogDebug` trough all the process

**Comment by aonez** [comment]
If the changes are ok for you, the case tests call to `[archive firstVolumePath]` should be replaced to `[archive filename]` and `[archive fileURL]`.

File: Classes/URKArchive.mm (hunk 13 above)
Comment (abbeycode) [reply Classes/URKArchive.mm]: No, all logging is for the dev to see, not the user.

Comment (abbeycode) [reply Classes/URKArchive.mm]: Also, look at the README for logging suggestions. This should probably be DEBUG level

File: Classes/URKArchive.mm (hunk 12 above)
Comment (abbeycode) [reply Classes/URKArchive.mm]: I have a few thoughts on this.

1. Since we're working with regular expressions in the `firstVolume...` method, it makes sense for it to use a path. Let it be true to itself
2. I think it would be clearer if `firstVolumePath` were a class method that takes a path, since it isn't using any internal state of the class
3. It only needs to be private. We can get rid of the public `firstVolumeURL` and `firstPathURL` instance methods, since all instances are going to represent the first volume. Basically, it will always be true that `firstVolumeURL == fileURL`

File: Classes/URKArchive.mm [hunk 18]
```diff
@@ -385,23 +395,75 @@ + (BOOL)urlIsARAR:(NSURL *)fileURL
...
+    
+    __block NSString *volumePath = fileURL.path;
```
Comment (abbeycode) [inline Classes/URKArchive.mm]: This doesn't need to be a `__block` variable, since it's not modified from inside a block. Also, once we refactor to a single class method (`+[URKArchive firstVolumePath:]`), it'll be the input argument.

File: Classes/URKArchive.mm [hunk 19]
```diff
@@ -385,23 +395,75 @@ + (BOOL)urlIsARAR:(NSURL *)fileURL
...
+    NSTextCheckingResult * match;
+    
+    if (volumePath.length)
```
Comment (abbeycode) [inline Classes/URKArchive.mm]: Invert this `if` and exit early. If `volumePath` is empty, log an error and return nil. That way we un-nest everything below, and you don't need to check for `match` twice. You can simplify the rest of the method's conditions.

File: Classes/URKArchive.mm (hunk 15 above)
Comment (abbeycode) [reply Classes/URKArchive.mm]: The `NSMutableOrderedSet` class doesn't sort members as they're added. You're better off using a regular `NSMutableSet` and sorting it when you convert to an array.

**Comment by aonez** [comment]
Go ahead with the pull. I think now all the elements are working, it's just a matter of "coding" preferences and optimisations that maybe you'll do faster. I'll take a look at the final code once is there 🤓 Obviously if you need anything from my part just ask.

**Comment by abbeycode** [comment]
@aonez Thanks again for all your work on this! I'll mention you in the PR I create to merge the `multivolume` branch into my release branch (`v2.9`).

**Comment by abbeycode** [comment]
@aonez Take a look at the PR I created (linked above) and let me know what you think. If it looks good, I'll merge to v2.9 and release a second beta soon.
